cli/pkg/validate/refs.go

```diff
@@ -25,6 +28,67 @@ func (rv *RefValidator) Validate(dc *catalog.DataCatalog) []ValidationError {
 			)
 		}
 	}
+
+	// Validate custom type references
+	for group, customTypes := range dc.CustomTypes {
+		for _, customType := range customTypes {
+			// Only object types with properties need validation
+			if customType.Type == "object" && len(customType.Properties) > 0 {
+				reference := fmt.Sprintf("#/custom-types/%s/%s", group, customType.LocalID)
+
+				for i, prop := range customType.Properties {
+
+					// Validate property reference format
+					matches := catalog.PropRegex.FindStringSubmatch(prop.ID)
+					if len(matches) != 3 {
+						errs = append(errs, ValidationError{
+							Reference: reference,
+							error:     fmt.Errorf("property reference at index %d has invalid format. Should be '#/properties/<group>/<id>'", i),
+						})
+						continue
+					}
+
+					// Validate property existence
+					groupName, propID := matches[1], matches[2]
+					if property := dc.Property(groupName, propID); property == nil {
+						errs = append(errs, ValidationError{
+							Reference: reference,
+							error:     fmt.Errorf("property reference '%s' at index %d not found in catalog", prop.ID, i),
+						})
+					}
+				}
+			}
+		}
+	}
+
+	// Validate property type custom type references
+	for group, props := range dc.Properties {
+		for _, prop := range props {
+			reference := fmt.Sprintf("#/properties/%s/%s", group, prop.LocalID)
+
+			// Check if the type field contains a custom type reference
+			if strings.HasPrefix(prop.Type, "#/custom-types/") {
+				matches := catalog.CustomTypeRegex.FindStringSubmatch(prop.Type)
+				if len(matches) != 3 {
+					errs = append(errs, ValidationError{
+						Reference: reference,
+						error:     fmt.Errorf("custom type reference in type field has invalid format. Should be '#/custom-types/<group>/<id>'"),
+					})
+					continue
+				}
+
+				// Validate custom type existence
+				customTypeGroup, customTypeID := matches[1], matches[2]
+				if customType := dc.CustomType(customTypeGroup, customTypeID); customType == nil {
+					errs = append(errs, ValidationError{
+						Reference: reference,
+						error:     fmt.Errorf("custom type reference '%s' not found in catalog", prop.Type),
+					})
+				}
+			}
+		}
+	}
```

Comment:
This is mostly a mental note, doesn't necessarily require any changes yet:
Looking for valid references (valid reference string, existence of referred resources) looks like a common pattern, which ideally could be abstracted to:
- handle common parsing logic, and resource lookup.
- have common consistent error reporting.

Lookup should be possible even at the YAML or generic resource representation, not necessarily depend on the parsed local catalog state the provider maintains. That would require significant changes though, that's why I'm not expecting any changes atm. It's important to keep in mind though because we're probably going to have to support cross-domain references e.g retl sql models to an account, or source to a tracking plan.

Reply:
Yes, you are right. This is why the current validation logic is flawed and limited. There are common validations check which doesn't need complete parsing of the files.